go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -2114,6 +2134,62 @@ func commonFlows(bridge *bridgeConfiguration) ([]string, error) {
 	return dftFlows, nil
 }
 
+// ovnToHostNetworkNormalActionFlows returns the flows that allow IP{v4,v6} traffic from the OVN network to the host network
+// when the destination is on the same node as the sender. This is necessary for pods in the default network to reach
+// localnet pods on the same node.
+func ovnToHostNetworkNormalActionFlows(netConfig *bridgeUDNConfiguration, srcMAC, dstIP string) []string {
+	var inPort, ctMark, ipFamily, ipFamilyDest string
+	var flows []string
+	var isIPv6 bool
+
+	if config.Gateway.Mode == config.GatewayModeShared {
+		inPort = netConfig.ofPortPatch
+		ctMark = netConfig.masqCTMark
+	} else if config.Gateway.Mode == config.GatewayModeLocal {
+		inPort = "LOCAL"
+		ctMark = ctMarkHost
+	} else {
+		return nil
+	}
+
+	if utilnet.IsIPv6CIDRString(dstIP) || utilnet.IsIPv6String(dstIP) {
+		isIPv6 = true
+		ipFamily = "ipv6"
+		ipFamilyDest = "ipv6_dst"
+	} else {
+		ipFamily = "ip"
+		ipFamilyDest = "nw_dst"
+	}
+
+	// IP traffic from the OVN network to the host network should be handled normally by the bridge instead of
+	// being output directly to the NIC by the existing flow at prio=100.
+	flows = append(flows,
+		fmt.Sprintf("cookie=%s, priority=102, in_port=%s, dl_src=%s, %s, %s=%s, "+
+			"actions=ct(commit, zone=%d, exec(set_field:%s->ct_mark)), output:NORMAL",
+			defaultOpenFlowCookie,
+			inPort,
+			srcMAC,
+			ipFamily,
+			ipFamilyDest,
+			dstIP,
+			config.Default.ConntrackZone,
+			ctMark))
+	if isIPv6 {
+		// Neighbor discovery in IPv6 happens through ICMPv6 messages to a special destination (ff02::1:ff00:0/104),
+		// which has nothing to do with the host subnets we're matching against in the flow above at prio=102.
+		// Let's allow neighbor discovery by matching against icmp type and in_port.
+		for _, icmpType := range []int{types.NeighborSolicitationICMPType, types.NeighborAdvertisementICMPType} {
+			flows = append(flows,
+				fmt.Sprintf("cookie=%s, priority=102, in_port=%s, dl_src=%s, icmp6, icmpv6_type=%d, "+
```

Comment:
so this function is going to be called on potentially multiple IP addresses in a for loop, but this particular flow is agnostic to the IP, so the same flow will be appended multiple times. Maybe we should move this out of this function and put it into the commonFlows itself under the ipv6 part.

Reply:
Ops. Good catch. I've reworked the function a little bit to pass the host cidrs and do everything inside the function instead of splitting it across different spots in the source code.